File: 백준/Silver/9012. 괄호/괄호.java
import java.util.*;

public class Main{
    public static void main(String[] args){
       Scanner sc=new Scanner(System.in);

        int T=sc.nextInt();
        sc.nextLine();

        for(int i=0;i<T;i++){
            Stack<Character> stack=new Stack<>();

            String input=sc.nextLine();
            boolean t=true;

            for(int j=0;j<input.length();j++){
                char c=input.charAt(j);

                if(c=='('){
                    stack.push(c);
                }else{
                    if(stack.empty()){
                        t=false;
                        break;
                    }else{
                        stack.pop();
                    }
                }
            }

            if(t && stack.empty()){
                System.out.println("YES");
            }else{
                System.out.println("NO");
            }
        }
    }
}

File: 백준/Silver/9012. 괄호/괄호.py
T=int(input())

for _ in range(T):
    arr=input()
    stack=[]
    is_valid = True
    
    for ch in arr:
        if ch=='(':
            stack.append(ch)
        else: # ')'
            if not stack:
                is_valid = False
                break  # 현재 테스트 케이스는 더 이상 볼 필요 없음
            
           
            if stack[-1]=='(':
                stack.pop()

        
    if is_valid and not stack:
        print("YES")
    else:
        print("NO")